morphia/src/main/java/org/mongodb/morphia/DatastoreImpl.java

```diff
@@ -1316,22 +1316,39 @@ private long nextValue(final Long oldVersion) {
         return dbObject;
     }
 
-    private <T> UpdateResults update(final Query<T> query, final UpdateOperations ops, final boolean createIfMissing, final boolean multi) {
-        return update(query, ops, createIfMissing, multi, getWriteConcern(query.getEntityClass()));
-    }
+    @Override
+    public <T> UpdateResults update(final Query<T> query, final UpdateOperations<T> operations, final UpdateOptions options) {
+        DBCollection dbColl = query.getCollection();
+        // TODO remove this after testing.
+        if (dbColl == null) {
+            dbColl = getCollection(query.getEntityClass());
+        }
 
-    @SuppressWarnings("rawtypes")
-    private <T> UpdateResults update(final Query<T> query, final UpdateOperations ops, final boolean createIfMissing, final boolean multi,
-                                     final WriteConcern wc) {
-        final DBObject u = ((UpdateOpsImpl) ops).getOps();
-        if (((UpdateOpsImpl) ops).isIsolated()) {
-            final Query<T> q = query.cloneQuery();
-            q.disableValidation().filter("$atomic", true);
-            return update(q, u, createIfMissing, multi, wc);
+        final MappedClass mc = getMapper().getMappedClass(query.getEntityClass());
+        final List<MappedField> fields = mc.getFieldsAnnotatedWith(Version.class);
+
+        DBObject queryObject = query.getQueryObject();
+        if (operations.isIsolated()) {
+            queryObject.put("$isolated", true);
```

Comment:
Line 1332 is also not covered bty any TestDatastore tests.  A test should be added at the least.